### PR TITLE
Add talks 1 to implement 1

### LINKED ISSUE #52: There should be a talk model
**Describe the feature or change you'd like**
Currently there is no talk model implemented in the reducer. This should be implemented with the field specified in the talks prototype.

**Additional context**
Ref: https://github.com/snek-at/prototype-talks

### LINKED ISSUE #53: There should be a talk util
**Describe the feature or change you'd like**
Currently there is no util to fetch talks. This should be implemented with the information provided by the prototype.

**Additional context**
Ref: https://github.com/snek-at/prototype-talks

### LINKED ISSUE #54: There should be a upload util
**Describe the feature or change you'd like**
Currently there is no file upload functionality provided by the intel. We should implement an upload functionality for talks to a specific file hosting service. Maybe anonfiles?

**Additional context**
Ref: https://github.com/snek-at/prototype-talks
Ref: https://anonfiles.com

### PR DESCRIPTION
**Pull request by schettn**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently there is no talk model in the database. #52
- Currently there is no way to query talks. #53
- Currently there is no upload functionality. #54

**What is the new behavior (if this is a feature change)?**
- Now there is a talk model with a reducer helper function for displaying. Closes #52
- Now there is a functionality to query talks by providing e.g Github usernames. Closes #53 
- Now there is an upload functionality. Currently this is adapted to talks. Closes #54 

**Other information**:

As written in the coding guidelines snek-tonic, no more points are set at JSDoc headers. Exceptions include cases in which a comment spans several lines.

Over time, the intel should be adapted to the new coding guide line. #55

- Ref: https://github.com/snek-at/tonic/blob/master/STYLE_GUIDE.md
- Ref: https://anonfiles.com
- Ref: https://github.com/snek-at/prototype-talks
- Ref: 🐍

### CODE REVIEW

File: src/reducer/database/models.ts [hunk 1]
```diff
@@ -111,6 +111,17 @@ interface IContribution {
   type: string;
   calendarId: number;
 }
+
+/** @interface Talk defines the structure of the contribution model. */
```
Comment (schettn) [inline src/reducer/database/models.ts:115]: Remove dot.
Ref: #55

File: src/reducer/database/models.ts [hunk 2]
```diff
@@ -960,6 +971,79 @@ class Contribution extends osm.models.ContributionSO implements IContribution {
    */
   save() {}
 }
+
+/**
```
Comment (schettn) [inline src/reducer/database/models.ts:975]: Remove dots.
Ref: #55

File: src/reducer/database/models.ts [hunk 3]
```diff
@@ -960,6 +971,79 @@ class Contribution extends osm.models.ContributionSO implements IContribution {
...
+ * @class A OSM model for a contribution object.
+ * @extends osm.models.ContributionSO Statement object.
```
Comment (schettn) [inline src/reducer/database/models.ts:977]: Change ContributionSO to TalkSO

File: src/reducer/database/models.ts [hunk 4]
```diff
@@ -960,6 +971,79 @@ class Contribution extends osm.models.ContributionSO implements IContribution {
    */
   save() {}
 }
+
+/**
+ * @class A OSM model for a contribution object.
```
Comment (schettn) [inline src/reducer/database/models.ts:976]: Change contribution to talk

File: src/reducer/database/models.ts [hunk 5]
```diff
@@ -960,6 +971,79 @@ class Contribution extends osm.models.ContributionSO implements IContribution {
...
+          "therefore an empty repository is provided!"
+      );
+
+      const repositoryData: any = {};
+
+      return new Repository(repositoryData);
+    }
+  }
+
+  /**
```
Comment (schettn) [inline src/reducer/database/models.ts:1041]: Remove dots.
Ref: #55

File: src/reducer/database/osm/models.ts [hunk 6]
```diff
@@ -1514,6 +1515,86 @@ abstract class ContributionSO extends BaseSO {
   //   }
   // }
 }
+
+/**
+ * @abstract
+ * @class A statement object which implements the talk sql statements.
```
Comment (schettn) [inline src/reducer/database/osm/models.ts:1521]: Remove dot:
Ref: #55

File: src/reducer/helper/talks.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+//> Models
+// Contains the repository and talk model of the database
+import { Repository, Talk } from "../database/models";
+//#endregion
+
+//#region > Interfaces
+/** @interface Talk defines the structure of a talk object */
+interface ITalk extends Talk {
+  /**
+   * Repository: The repository object which contains the talk.
```
Comment (schettn) [inline src/reducer/helper/talks.ts:11]: Remove dot:
Ref: #55

File: src/reducer/index.ts [hunk 8]
```diff
@@ -10,18 +10,33 @@ import * as helper from "./helper";
 //#region > Classes
 /** @class Reduce the data of the models. */
```
Comment (schettn) [inline src/reducer/index.ts:11]: Remove dot:
Ref: #55

File: src/utils/talks/index.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+// Contains all provided converters for Github Talk data
+import * as converter from "./converter";
+//#endregion
+
+//#region > Interfaces
+/**
+ * @interface Source defines the structure of a source object which is required
+ *                   to query talks.
+ */
+interface ISource {
```
Comment (schettn) [inline src/utils/talks/index.ts:19]: Remove dots:
Ref: #55

File: src/utils/talks/converter.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+//#endregion
+
+//#region > Functions
+/**
+ * @function
+ * @param rawData
+ * @description Converter for talk data from the Github APIv3
+ */
+async function run(rawData: any) {
+  console.log(rawData);
```
Comment (Aichnerc) [inline src/utils/talks/converter.ts:14]: Is this console.log required?

Comment (schettn) [reply src/utils/talks/converter.ts:14]: Right. I will remove it together with a bugfix.

File: src/utils/talks/converter.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+      downloadUrl = item.html_url.replace("/blob/", "/raw/");
+      displayUrl += downloadUrl;
+    }
+
+    const talk = {
+      name: item.name,
+      url: item.html_url,
+      displayUrl,
+      downloadUrl,
+      path: item.path
```
Comment (pinterid) [inline src/utils/talks/converter.ts:31]: Missing comma.

File: src/utils/talks/converter.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+    const repository = {
+      name: item.repository.name,
+      url: item.repository.html_url,
+      avatarUrl: item.repository.owner.avatar_url,
+      owner: {
+        avatarUrl: item.repository.owner.avatar_url,
+        url: item.repository.owner.url,
+        fullname: item.repository.owner.fullname,
+        username: item.repository.owner.username
```
Comment (pinterid) [inline src/utils/talks/converter.ts:42]: Missing comma.

File: src/utils/talks/index.ts [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+ */
+async function generate(source: ISource) {
+  const client = new WebClient("https://api.github.com/", {
+    authorization: source.authorization,
+  });
+
+  const run = async (name: string, user: boolean = true) => {
+    /* Page Number */
+    let pn = 1;
+
```
Comment (pinterid) [inline src/utils/talks/index.ts:51]: Is this newline intended?

File: src/utils/talks/index.ts [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+      let response = await client.scraper.getJson<{
+        items: [];
+        total_count: 0;
+      }>(path);
+
+      totalItems = response.total_count;
+      currentItems += response.items.length;
+
+      await converter.run({ talks: response.items });
```
Comment (pinterid) [inline src/utils/talks/index.ts:74]: Could add a newline after converter.run.

File: src/utils/upload/anonfiles.ts [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  };
+}
+//#endregion
+
+//#region > Functions
+/**
+ * @function
+ * @param {Document} dom A specific download site of anonfile
+ * @returns {string | null} The download URL if available
+ * @description Extracts the download url from a download site
```
Comment (pinterid) [inline src/utils/upload/anonfiles.ts:35]: Change "url" to "URL"

File: src/utils/talks/paths.ts [hunk 16]
```diff
@@ -0,0 +1,18 @@
+//#region > Paths
+/** Provides the Github APIv3 path for querying user talks */
```
Comment (kleberbaum) [inline src/utils/talks/paths.ts:2]: Please add a periods.

File: src/utils/talks/paths.ts [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+const user = (username: string, page: number) =>
+  `/search/code?q=user:${username} in:file extension:pdf&page=${page}`;
+
+/** Provides the Github APIv3 path for querying organization talks */
```
Comment (kleberbaum) [inline src/utils/talks/paths.ts:6]: Please add a periods.

File: src/utils/upload/anonfiles.ts [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+//## npm install snek-client
+// Contains a client for web calls
+import { WebClient } from "snek-client";
+//#endregion
+
+//#region > Interfaces
+/** 
+ * @todo Convert this to a generic file interface
+ * @interface Talk defines the structure of a not converted talk object
+ */
```
Comment (kleberbaum) [inline src/utils/upload/anonfiles.ts:11]: Please add a periods.

File: src/utils/upload/anonfiles.ts [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+interface ITalk {
+  name: string;
+  displayUrl: string | null;
+  downloadUrl: string | null;
+  html_url: string;
+  path: string;
+  /** @todo Remove repository from the interface */
```
Comment (kleberbaum) [inline src/utils/upload/anonfiles.ts:18]: Please add a periods.

File: src/utils/upload/anonfiles.ts [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+}
+//#endregion
+
+//#region > Functions
+/**
+ * @function
+ * @param {Document} dom A specific download site of anonfile
+ * @returns {string | null} The download URL if available
+ * @description Extracts the download url from a download site
+ */
```
Comment (kleberbaum) [inline src/utils/upload/anonfiles.ts:36]: Please add a periods.

File: src/utils/upload/anonfiles.ts [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+function getDownloadUrl(dom: Document): string | null {
+  return dom.getElementsByTagName("a")[1].getAttribute("href");
+}
+
+/**
+ * @function
+ * @param {Blob} file A file to be uploaded
+ * @returns {ITalk} A unconverted talk object
+ * @description Upload a file to anonfile and return 
+ */
```
Comment (kleberbaum) [inline src/utils/upload/anonfiles.ts:46]: Please add a periods.

File: src/utils/talks/index.ts (hunk 13 above)
Comment (schettn) [reply src/utils/talks/index.ts:51]: Nope. I will remove it.

**Comment by schettn** [comment]
I just removed and/or resolved all comments which includes "Please add a period.", because they were not valid at this point. 
I know that these comments were created to maintain persistence, but a PR will be opened in the next few days to deal with this persistence.
LG me 🧷

**Comment by pinterid** [comment]
Add a function that gets the download URL from Annofiles before merging.

**Pull request merged**